text/0000-PXE-Booting-UEFI.md

```diff
@@ -0,0 +1,228 @@
+both BIOS and UEFI naming conventions (grub/menu.lst vs grub/efidefault and
+grub/menu.lst.01aabbccddeeff vs grub/01-AA-BB-CC-DD-EE-FF). Also, Pxegrub
+variant will use new base directory "grub/" instead of "boot/grub/" for
+consistency with "grub2" and "pxelinux.cfg/".
+
+### [Foreman core #12634](http://projects.theforeman.org/issues/12634) - New host/hostgroup flags pxe_template and loader
+
+Host and hostgroups must be enhanced with two new flags: PXE template specifies
+PXE template kind and variant pair to use. Until today, this was hardcoded to
+PXELinux or PXEGrub by selected operating system. Optionally, template variant
+flag from OS could be simply dropped because PXE template implies variant to
+use and there are no variants which share the same configuration syntax (this
+is an open item).
+
+Orchestration TFTP code must be modified to use multiple "KIND LOADER default
+local boot" and "KIND LOADER global default" templates. Local and global boot
+templates must be added for Grub2 new kind. Naming convention will be:
+
+* PXELinux BIOS default local boot
+* PXELinux UEFI default local boot
+* PXEGrub BIOS default local boot
+* PXEGrub UEFI default local boot
+* etc. (the same for global default templates)
+
+But only kind/variant is not enough, users need to select BIOS or UEFI booting
+method. This is where Loader flag comes in - selection of three constants:
+BIOS, UEFI and SecureBoot UEFI. The first two are obvious causing selection of
+the correct DHCP filename when deploying TFTP configuration, the third one
+select "shim.efi" chain loading capability for secure booting.
+
+Similarly to root password, a global setting could be provided for one or both
+flags for missing values (open item).
+
+### [Foreman installer #14920](http://projects.theforeman.org/issues/14920) - Change dhcpd.conf to support EFI loaders
+
+Default DHCPD configuration must be changed in order to allow booting unknown
+UEFI hosts as it was shown above (dhcpd.conf snippet). Using shim.efi is
+perhaps better default than loading Grub directly.
+
+### [Foreman installer #12635](http://projects.theforeman.org/issues/12635) - Options to deploy Grub and PXELinux EFI loaders in TFTP root
+
+It looks like Grub2 is the best and most compatible way of PXE booting EFI
+hardware and it is available in most Linux distributions. It makes sense to
+provide support in our Puppet classes and installer to deploy these.
+Instructions on where to find the firmware in RHEL7 are at
+[access.redhat.com](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Installation_Guide/chap-installation-server-setup.html)
+(RHEL6 does not ship Grub2 so installer should fail or skip that). Similarly
+for Debian and Ubuntu it is available for [direct
+download](http://archive.ubuntu.com/ubuntu/dists/trusty/main/uefi/grub2-amd64/current/)
+or in debian-installer-8-netboot-amd64 package.
+
+UEFI shim shall be installed if possible. Expected file paths are defined in
```

Comment:
Not all distributions uses Shim approach, for example Debian family signs Grub itself and they distribute two copies of the same. https://wiki.ubuntu.com/UEFI/PXE-netboot-install